Lombiq.Tests.UI/Extensions/VisualVerificationUITestContextExtensions.cs

```diff
@@ -0,0 +1,509 @@
+    public static void VisualAssert(
+        this UITestContext context,
+        IWebElement element,
+        Bitmap reference,
+        double meanErrorPercentageThreshold,
+        Rectangle? roi = null,
+        Action<VisualMatchConfiguration> configurator = null) => context
+        .VisualAssert(
+            element,
+            reference,
+            diff =>
+                AssertInternal(
+                    (actual, expected) => actual <= expected,
+                    diff.PixelErrorPercentage,
+                    meanErrorPercentageThreshold,
+                    nameof(diff.PixelErrorPercentage),
+                    ConditionLessThenOrEqual),
+            roi,
+            configurator);
+
+    /// <summary>
+    /// Compares the reference image and screenshot of the element given by <paramref name="elementSelector"/>.
+    /// </summary>
+    /// <param name="context">The <see cref="UITestContext"/> in which the extension is executed on.</param>
+    /// <param name="elementSelector">Selector for the target element.</param>
+    /// <param name="reference">Reference image.</param>
+    /// <param name="comparator">To validate the comparison result.</param>
+    /// <param name="roi">Region of interest. Can be  null.</param>
+    /// <param name="configurator">Action callback to configure the behavior. Can be null.</param>
+    private static void VisualAssert(
+        this UITestContext context,
+        By elementSelector,
+        Bitmap reference,
+        Action<ICompareResult> comparator,
+        Rectangle? roi = null,
+        Action<VisualMatchConfiguration> configurator = null) => context
+        .VisualAssert(context.Get(elementSelector), reference, comparator, roi, configurator);
+
+    /// <summary>
+    /// Compares the reference image and screenshot of the element.
+    /// </summary>
+    /// <param name="context">The <see cref="UITestContext"/> in which the extension is executed on.</param>
+    /// <param name="element">Target element.</param>
+    /// <param name="reference">Reference image.</param>
+    /// <param name="comparator">To validate the comparison result.</param>
+    /// <param name="roi">Region of interest. Can be  null.</param>
+    /// <param name="configurator">Action callback to configure the behavior. Can be null.</param>
+    private static void VisualAssert(
+        this UITestContext context,
+        IWebElement element,
+        Bitmap reference,
+        Action<ICompareResult> comparator,
+        Rectangle? roi = null,
+        Action<VisualMatchConfiguration> configurator = null)
+    {
+        var configuration = new VisualMatchConfiguration();
+        configurator?.Invoke(configuration);
+
+        var cropRegion = roi ?? new Rectangle(0, 0, reference.Width, reference.Height);
+
+        // We take a screenshot and append it to the failure dump.
+        var fullScreenImage = context.TakeScreenshot()
+            .ToBitmap();
+        context.AppendFailureDump(
```

Comment:
These failure dump appends should only happen if the verification failed.

Reply:
Yes exactly.

Reply:
I mean, currently they happen any time `AssertVisualVerification()` is invoked.

Reply:
When the test fails on something else, they'll be part of the failure dump despite verification test passing, no? Then I'd especially move this to a `catch`.

Reply:
I found a better solution. I introduced a new argument `Type[] inCaseOf` for `AppendFailureDump`, which filters out that `FailureDumpItem` which does not match any given exception type.
With this change, the code remains readable, I think.
Conditions:
- if `inCaseOf is null`: the item will be dumped in all cases.
- if `inCaseOf` is an empty `Array`: the item will never be dumped.
- if `inCaseOf` contains the type of exception thrown, the item will be dumped.

With this change, the code remains readable, I think.
What do you think?

```csharp
        context.AppendFailureDump(
            Path.Combine(
                VisualVerificationMatchNames.DumpFolderName,
                new[]
                {
                    configuration.FileNamePrefix,
                    VisualVerificationMatchNames.CroppedElementImageFileName,
                    configuration.FileNameSuffix,
                }
                .JoinNotEmptySafe("-")),
            elementImage.Clone(),
            messageIfExists: HintFailureDumpItemAlreadyExists,
            inCaseOf: new[] { typeof(VisualVerificationAssertionException) });
```

Reply:
I don't really understand why this is necessary. In this method, `comparator(diff);` at the end will throw an exception if the assertion failed. Why can't we append to the failure dump only when catching (and immediately rethrowing) this exception?

The approach you show is quite cool, but I don't see why we need do defer this check to the end, when creating the failure dump, vs not adding the items to the failure dump in the first place.